### PR TITLE
bpf: datapath: Rewite base devices setup in Go

### PR DESCRIPTION
**Pull request by vadorovsky**
This change removes the part of init.sh responsible for base devices
setup and remplements it in Go.

Ref: #920

Signed-off-by: Michal Rostecki <mrostecki@opensuse.org>

### CODE REVIEW

**Comment by mrostecki** [comment]
test-me-please

**Comment by pchaigno** [comment]
test-me-please

**Comment by mrostecki** [comment]
This failure is legit and I'll try to fix it tomorrow:

```
23:24:37  RuntimePolicies
23:24:37  /home/jenkins/workspace/Cilium-PR-Runtime-4.9/runtime-gopath/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:478
23:24:37    Init Policy Test
23:24:37    /home/jenkins/workspace/Cilium-PR-Runtime-4.9/runtime-gopath/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:478
23:24:37      Init Egress Policy Test [AfterEach]
23:24:37      /home/jenkins/workspace/Cilium-PR-Runtime-4.9/runtime-gopath/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:514
23:24:37  
23:24:37      Found 1 Cilium logs matching list of errors that must be investigated:
[2020-11-05T22:24:37.386Z]     empty argument passed to bpf/init.sh
23:24:37  
23:24:37      /home/jenkins/workspace/Cilium-PR-Runtime-4.9/runtime-gopath/src/github.com/cilium/cilium/test/ginkgo-ext/scopes.go:421
```

But some others are failing with:

```
22:38:08  Error response from daemon: toomanyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit
```

**Comment by mrostecki** [comment]
test-me-please

**Comment by mrostecki** [comment]
retest-4.19

**Comment by mrostecki** [comment]
test-gke

File: pkg/datapath/loader/netlink.go [hunk 1]
```diff
@@ -151,3 +155,164 @@ func RemoveTCFilters(ifName string, tcDir uint32) error {
...
+			{Name: fmt.Sprintf("net.ipv4.conf.%s.rp_filter", ifName), Val: "0", IgnoreErr: false},
+			{Name: fmt.Sprintf("net.ipv4.conf.%s.accept_local", ifName), Val: "1", IgnoreErr: false},
+			{Name: fmt.Sprintf("net.ipv4.conf.%s.send_redirects", ifName), Val: "0", IgnoreErr: false},
+		}...)
+	}
+	if err := sysctl.ApplySettings(sysSettings); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (aanm) [inline pkg/datapath/loader/netlink.go:191]: `return sysctl.ApplySettings(sysSettings)`

File: pkg/datapath/loader/netlink.go [hunk 2]
```diff
@@ -151,3 +155,164 @@ func RemoveTCFilters(ifName string, tcDir uint32) error {
...
+	}
+	peer, err := netlink.LinkByName(peerName)
+	if err != nil {
+		return err
+	}
+	if err := setupDev(peer); err != nil {
+		return err
+	}
+
+	return nil
```
Comment (aanm) [inline pkg/datapath/loader/netlink.go:243]: `return setupDev(peer)`

File: pkg/datapath/loader/netlink.go [hunk 3]
```diff
@@ -151,3 +155,164 @@ func RemoveTCFilters(ifName string, tcDir uint32) error {
...
+}
+
+// setupBaseDevice decides which and what kind of interfaces should be set up as
+// the first step of datapath initialization, then performs the setup (and
+// creation, if needed) of those interfaces. It returns two links and an error.
+// By default, it sets up the veth pair - cilium_host and cilium_net.
+// In flannel mode, it sets up the native interface used by flannel.
+// In ipvlan mode, it creates the cilium_host ipvlan with the native device as a
+// parent.
+func setupBaseDevice(nativeDevs []netlink.Link, mode string, mtu int) (netlink.Link, netlink.Link, error) {
```
Comment (aanm) [inline pkg/datapath/loader/netlink.go]: I would suggest to replace the `mode string` with `mode baseDeviceMode` where `baseDeviceMode` is `type baseDeviceMode string`

**Review by aanm — APPROVED** [review]
Only small nits, overall LGTM

File: pkg/datapath/loader/netlink.go [hunk 4]
```diff
@@ -151,3 +155,164 @@ func RemoveTCFilters(ifName string, tcDir uint32) error {
...
+	switch mode {
+	case "flannel":
+		if err := setupDevs(nativeDevs...); err != nil {
+			return nil, nil, err
+		}
+		return nativeDevs[0], nativeDevs[0], nil
+	case "ipvlan":
+		ipvlan, err := setupIpvlan("cilium_host", nativeDevs[0])
```
Comment (pchaigno) [inline pkg/datapath/loader/netlink.go]:
```suggestion
		ipvlan, err := setupIpvlan(defaults.HostDevice, nativeDevs[0])
```

File: pkg/datapath/loader/netlink.go [hunk 5]
```diff
@@ -151,3 +155,164 @@ func RemoveTCFilters(ifName string, tcDir uint32) error {
...
+	default:
+		if err := setupVethPair("cilium_host", "cilium_net"); err != nil {
+			return nil, nil, err
+		}
+
+		linkHost, err := netlink.LinkByName("cilium_host")
+		if err != nil {
+			return nil, nil, err
+		}
+		linkNet, err := netlink.LinkByName("cilium_net")
```
Comment (pchaigno) [inline pkg/datapath/loader/netlink.go]:
```suggestion
		linkNet, err := netlink.LinkByName(defaults.SecondHostDevice)
```

**Comment by mrostecki** [comment]
test-me-please

**Review by pchaigno — APPROVED** [review]
Looks good to me. One minor nit below.

Thanks for tackling this @mrostecki! :tada:

**Comment by mrostecki** [comment]
test-me-please

**Comment by brb** [comment]
@mrostecki Just echoing https://github.com/cilium/cilium/pull/14043#issuecomment-729892492. Have you validated ipvlan setup?

**Comment by mrostecki** [comment]
@brb I didn't. I will do that later today or on Monday.

**Comment by mrostecki** [comment]
ipvlan doesn't work, I'm working on the fix. Sorry for that!

**Pull request merged**